File: surf-cloud-api/src/main/java/dev/slne/surf/cloud/api/netty/NettySource.java
package dev.slne.surf.cloud.api.netty;

import dev.slne.surf.cloud.api.netty.packet.NettyPacket;
import lombok.Getter;
import org.apache.commons.lang3.builder.ToStringBuilder;

public class NettySource {

  @Getter
  private final NettyBase nettyBase;

  NettySource(NettyBase nettyBase) {
    this.nettyBase = nettyBase;
  }

  public void sendPacket(NettyPacket<?> packet) {
    // send packet
  }

  @Override
  public String toString() {
    return ToStringBuilder.reflectionToString(this);
  }
}

File: surf-cloud-api/src/main/kotlin/dev/slne/surf/cloud/api/netty/source/NettySource.kt
package dev.slne.surf.cloud.api.netty.source

import dev.slne.surf.cloud.api.netty.NettyBase
import dev.slne.surf.cloud.api.netty.annotation.PlatformRequirement
import dev.slne.surf.cloud.api.netty.packet.NettyPacket
import org.jetbrains.annotations.ApiStatus

@ApiStatus.NonExtendable
@PlatformRequirement(PlatformRequirement.Platform.COMMON)
interface NettySource<Client : ProxiedNettySource<Client>> {
    val base: NettyBase<Client>

    val proxied: Boolean
        get() = this is ProxiedNettySource<*>

    fun sendPacket(packet: NettyPacket<*>)
}
